src/IO/CachedInMemoryReadBufferFromFile.cpp

```diff
@@ -0,0 +1,178 @@
+
+        if (!chunk->isPrefixPopulated(chunk_size))
+        {
+            /// A few things could be improved here, which may or may not be worth the added complexity:
+            ///  * If the next file chunk is in cache, use in->setReadUntilPosition() to limit the read to
+            ///    just one chunk. More generally, look ahead in the cache to count how many next chunks
+            ///    need to be downloaded. (Up to some limit? And avoid changing `in`'s until-position if
+            ///    it's already reasonable; otherwise we'd increase it by one chunk every chunk, discarding
+            ///    a half-completed HTTP request every time.)
+            ///  * If only a subset of pages are missing from this chunk, download only them.
+            ///    (With some threshold for avoiding short seeks.)
+            ///    This is probably not actually useful in practice because our PageCache normally uses
+            ///    2 MiB transparent huge pages (because MADV_FREE is too slow without them), with
+            ///    just one OS page per chunk.
+            ///  * If our [position, read_until_position) covers only part of the chunk, we could download
+            ///    just that part. (Which would be bad if someone else needs the rest of the chunk and has
+            ///    to do a whole new HTTP request to get it. Unclear what the policy should be.)
+            ///  * Instead of doing in->next() in a loop until we get the whole chunk, we could return the
+            ///    results as soon as in->next() produces them.
+            ///    (But this would make the download_mutex situation much more complex, similar to the
+            ///    FileSegment::State::PARTIALLY_DOWNLOADED and FileSegment::setRemoteFileReader() stuff.)
+
+            Buffer prev_in_buffer = in->internalBuffer();
+            SCOPE_EXIT({ in->set(prev_in_buffer.begin(), prev_in_buffer.size()); });
+
+            in->set(chunk->getChunk()->data, chunk_size);
+            in->seek(cache_key.offset, SEEK_SET);
+            chassert(!in->available());
+
+            size_t pos = 0;
+            while (pos < chunk_size)
+            {
+                char * piece_start = chunk->getChunk()->data + pos;
+                size_t piece_size = chunk_size - pos;
+                in->set(piece_start, piece_size);
+
+                if (!in->next())
+                    throw Exception(ErrorCodes::UNEXPECTED_END_OF_FILE, "File {} ended after {} bytes, but we expected {}",
+                        getFileName(), cache_key.offset + pos, file_size.value());
+
+                chassert(in->position() >= piece_start && in->buffer().end() <= piece_start + piece_size);
+
+                size_t n = in->available();
+                if (in->position() != piece_start)
+                    memmove(piece_start, in->position(), n);
+                in->position() += n;
+                pos += n;
+            }
+
+            chunk->markPrefixPopulated(chunk_size);
```

Comment:
Hm, right, we could detect here if some prefix was already downloaded and `in->seek()` to that instead of beginning of the chunk. Added it to the "things could be improved" comment above, but not implementing it now because it seems unlikely to be worth the complexity (in particular, are short `in->seek()`s faster than reads? seek() implementations usually have logic to avoid short seeks: if the new offset is still within the working_buffer, just advance the position pointer; but that won't work if we're assigning the buffer from the outside as that clears working_buffer; so we'd need to either refactor how this external-buffer mode works or add a short-seek threshold here or something).